### ISSUE TITLE
Replace dynamic module API calls

### ISSUE THREAD

**Issue by x1B** [LaxarJS/laxar#304]
The switch to webpack (#298) was mostly successful, but there are some places in the code that have been using System.js (and before that, RequireJS) to perform dynamic loading of modules or dynamic resolution of paths.
For v2.0 we want to resolve all paths at build-time, eliminating the need for these calls.

For now, the corresponding code locations have been marked with `TODO (#304)` and contain some sort of workaround which allows us to continue, but will break down in many real-world situations.

This ticket is complete when all of these locations have been fixed (e.g. by using precomputed paths).

**Comment by x1B** [comment]
Implemented on `master` (`v2.0.0`).

### BREAKING CHANGE

This change will usually not affect LaxarJS project development directly (its effects are encapsulated in the tooling layer), *except* from the change in the `laxar.bootstrap` signature (see below).

The mechanism to get artifact resources and modules into the runtime was changed. Previously, two objects (`widgetModules` and `resources`) had to be passed into the runtime. The *widget modules* contained all widget and control implementation modules, grouped by integration technology. The *resources* were basically a file system snapshot containing everything that might be relevant for running in the browser (styles, widget descriptors etc.). In this old setup, artifact *references* had to be resolved into paths both when building and when running on the client. This means that we had to add fairly complicated means of resolving e.g. RequireJS paths into the client, which break down when the build tooling changes too much.

The new interface is much more abstract: The [build tooling](//github.com/LaxarJS/laxar-tooling) must produce a single JS *artifacts bundle module*, which is passed as the `artifacts` option to `laxar.bootstrap`. That bundle contains

 - assets and code for each artifact used within a flow (depending on the artifact type: descriptor, definition, module and other assets, such as template and styles),

 - a mapping of each *artifact reference* reachable from the flow to its assets and code.


This ensures that:

 - references need only be *resolved* at build time, and can simply be *looked up* at runtime,

 - no lookup-paths need to be configured for LaxarJS, except sometimes a `base` URL for all assets,

 - the full feature-set of webpack (or comparable tools) can be utilized to resolve and process assets,

 - all artifacts have exactly one canonical `name`, which is encoded into their `descriptor` by the build tools.

See [laxar-tooling#2](https://github.com/LaxarJS/laxar-tooling/issues/2) for the build side of this issue.

**Comment by x1B** [comment]
### Addition

To illustrate the format of the new artifacts bundle module, here is an example listing for the [shop-demo](http://laxarjs.github.io/shop-demo/). It has an additional (and useless) *ax-accordion-widget*, just for the bundle to contain a control. Note that all artifacts have a descriptor with a *name*, but for widgets and controls that descriptor is actually backed by a JSON file.

```js
/* eslint-env node */
/* eslint-disable max-len */
module.exports = {
   aliases: {
      flows: {
         main: 0
      },
      pages: {
         shop_demo: 0,
         finish_order: 1
      },
      layouts: {
         application: 0
      },
      themes: {
         'cube': 0,
         'default': 1
      },
      widgets: {
         'shop-demo/dummy-articles-activity': 0,
         'shop-demo/article-browser-widget': 1,
         'shop-demo/article-teaser-widget': 2,
         'shop-demo/shopping-cart-widget': 3,
         'shop-demo/article-search-box-widget': 4,
         'shop-demo/headline-widget': 5,
         'ax-accordion-widget': 6
      },
      controls: {
         'laxar-accordion-control': 0
      }
   },
   flows: [ {
      descriptor: { name: 'main' },
      definition: require( '!!json!./application/flows/main.json' )
   } ],
   pages: [ {
      descriptor: { name: 'shop_demo' },
      definition: require( '!!json!./application/pages/shop_demo.json' )
   }, {
      descriptor: { name: 'finish_order' },
      definition: require( '!!json!./application/pages/finish_order.json' )
   } ],
   layouts: [ {
      descriptor: { name: 'application' },
      assets: {
         'default.theme': {
            'application.html': {
               content: require( '!!raw!./application/layouts/application/default.theme/application.html' )
            },
            'css/application.css': {
               url: './application/layouts/application/default.theme/css/application.css'
            }
         }
      }
   } ],
   themes: [ {
      descriptor: { name: 'cube.theme' },
      assets: {
         'cube.theme': {
            'css/theme.css': {
               url: './includes/themes/cube.theme/css/theme.css'
            }
         }
      }
   }, {
      descriptor: { name: 'default.theme' },
      assets: {
         'default.theme': {
            'css/theme.css': {
               url: './includes/lib/laxar-uikit/themes/default.theme/css/theme.css'
            }
         }
      }
   } ],
   widgets: [ {
      descriptor: require( '!!json!./includes/widgets/shop-demo/dummy-articles-activity/widget.json' ),
      module: require( './includes/widgets/shop-demo/dummy-articles-activity/dummy-articles-activity' )
   }, {
      descriptor: require( '!!json!./includes/widgets/shop-demo/article-browser-widget/widget.json' ),
      module: require( './includes/widgets/shop-demo/article-browser-widget/article-browser-widget' ),
      assets: {
         'default.theme': {
            'article-browser-widget.html': {
               content: require( '!!raw!./includes/widgets/shop-demo/article-browser-widget/default.theme/article-browser-widget.html' )
            },
            'css/article-browser-widget.css': {
               url: './includes/widgets/shop-demo/article-browser-widget/default.theme/css/article-browser-widget.css'
            }
         },
         'cube.theme': {
            'css/article-browser-widget.css': {
               url: './includes/widgets/shop-demo/article-browser-widget/cube.theme/css/article-browser-widget.css'
            }
         }
      }
   }, {
      descriptor: require( '!!json!./includes/widgets/shop-demo/article-teaser-widget/widget.json' ),
      module: require( './includes/widgets/shop-demo/article-teaser-widget/article-teaser-widget' ),
      assets: {
         'default.theme': {
            'css/article-teaser-widget.css': {
               url: './includes/widgets/shop-demo/article-teaser-widget/default.theme/css/article-teaser-widget.css'
            }
         },
         'cube.theme': {
            'css/article-teaser-widget.css': {
               url: './includes/widgets/shop-demo/article-teaser-widget/cube.theme/css/article-teaser-widget.css'
            }
         }
      }
   }, {
      descriptor: require( '!!json!./includes/widgets/shop-demo/shopping-cart-widget/widget.json' ),
      module: require( './includes/widgets/shop-demo/shopping-cart-widget/shopping-cart-widget' ),
      assets: {
         'default.theme': {
            'shopping-cart-widget.html': {
               content: require( '!!raw!./includes/widgets/shop-demo/shopping-cart-widget/default.theme/shopping-cart-widget.html' )
            },
            'css/shopping-cart-widget.css': {
               url: './includes/widgets/shop-demo/shopping-cart-widget/default.theme/css/shopping-cart-widget.css'
            }
         },
         'cube.theme': {
            'css/shopping-cart-widget.css': {
               url: './includes/widgets/shop-demo/shopping-cart-widget/cube.theme/css/shopping-cart-widget.css'
            }
         }
      }
   }, {
      descriptor: require( '!!json!./includes/widgets/shop-demo/article-search-box-widget/widget.json' ),
      module: require( './includes/widgets/shop-demo/article-search-box-widget/article-search-box-widget' ),
      assets: {
         'default.theme': {
            'article-search-box-widget.html': {
               content: require( '!!raw!./includes/widgets/shop-demo/article-search-box-widget/default.theme/article-search-box-widget.html' )
            },
            'css/article-search-box-widget.css': {
               url: './includes/widgets/shop-demo/article-search-box-widget/default.theme/css/article-search-box-widget.css'
            }
         }
      }
   }, {
      descriptor: require( '!!json!./includes/widgets/shop-demo/headline-widget/widget.json' ),
      module: require( './includes/widgets/shop-demo/headline-widget/headline-widget' ),
      assets: {
         'default.theme': {
            'headline-widget.html': {
               content: require( '!!raw!./includes/widgets/shop-demo/headline-widget/default.theme/headline-widget.html' )
            }
         }
      }
   }, {
      descriptor: require( '!!json!./includes/widgets/ax-accordion-widget/widget.json' ),
      module: require( './includes/widgets/ax-accordion-widget/ax-accordion-widget' ),
      assets: {
         'default.theme': {
            'ax-accordion-widget.html': {
               content: require( '!!raw!./includes/widgets/ax-accordion-widget/default.theme/ax-accordion-widget.html' )
            },
            'css/article-search-box-widget.css': {
               url: './includes/widgets/ax-accordion-widget/default.theme/css/ax-accordion-widget.css'
            }
         }
      }
   } ],
   controls: [
      {
         descriptor: require( '!!json!./includes/controls/laxar-accordion-control/control.json' ),
         module: require( './includes/controls/laxar-accordion-control/ax-accordion-control' ),
         assets: {
            'default.theme': {
               'css/ax-accordion-control.css': {
                  url: './includes/widgets/ax-accordion-control/default.theme/css/ax-accordion-control.css'
               }
            }
         }
      }
   ]
};
```
